cql3/restrictions/statement_restrictions.cc

```diff
@@ -915,5 +915,28 @@ bool token_restriction::slice::is_satisfied_by(const schema& schema,
     return satisfied;
 }
 
+bool single_column_restriction::LIKE::is_satisfied_by(const schema& schema,
+        const partition_key& key,
+        const clustering_key_prefix& ckey,
+        const row& cells,
+        const query_options& options,
+        gc_clock::time_point now) const {
+    if (!_column_def.type->is_string()) {
+        throw exceptions::invalid_request_exception("LIKE is allowed only on string types");
+    }
+    auto cell_value = get_value(schema, key, ckey, cells, now);
+    return !cell_value ? false :
+            cell_value->with_linearized([&] (bytes_view bv) {
+                 return _like_matcher(bv);
+            });
+}
+
+bool single_column_restriction::LIKE::is_satisfied_by(bytes_view data, const query_options& options) const {
+    if (!_column_def.type->is_string()) {
+        throw exceptions::invalid_request_exception("LIKE is allowed only on string types");
+    }
+    return _like_matcher(data);
+}
```

Comment:
It will be terrible if we have to construct the _like_matcher here for every row (which is where you get the query_options for bind_and_get()). So I think we need something like single_column_restriction::per_query_state:

Beginning of query: call all the per_query_state and stash in a vector (LIKE will compute a like_matcher there and store it)
Each row: for each single_column_restriction, provide the matching per_query_state (LIKE will static_cast and extract the per query state)

The other s_c_rs will also benefit. @psarna please validate the idea (or invalidate it, as the case may be).

But let's start with correctness and do the optimization later.

Reply:
It's even more involved. The `term` used for the pattern could be a function, and function parameters could be columns (I think). So we can only do this if the matcher doesn't depend on any column (which is the common case, but we also have to be prepared for the uncommon case).